model/src/main/kotlin/Result.kt

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (C) 2017-2020 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package com.here.ort.model
+
+/**
+ * The result of an operation with the return type [R].
+ */
+sealed class Result<R>
```

Comment:
Have you considered using Kotlin's [Result class](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-result/) instead?

Reply:
Yes, but it's not ideal for this use case. The sealed class provides a much nicer API IMO, e.g. because the value field does not have to be nullable and it works well with `when` expressions.

Reply:
Thanks. Would you mind adding the information that Kotlin's Result has been considered and why it doesn't fit here to the commit message?

Reply:
Or maybe even to the class docs?

Reply:
Not saying your approach would be changed but out of interrest maybe you guys can give a hint on the following.
The sealed class approach seems to have also drawbacks. 
I wonder if code can't be written anymore nicely with just using `if` but one gets dragged into using
when which is a bit more verbose and adds a level of nesting. Can the return early pattern below still be written
without using `when` nicely? (I fear I'm missing something)

```
val result = storage.add(...)
if (!result.success) return emptyList()

val value = result.result!!
// do somthing with value
```

Reply:
> Can the return early pattern below still be written without using `when` nicely? (I fear I'm missing something)

Yes, you would use:

```kotlin
if (result is Success) ...
if (result is Failure) ...
```

Reply:
@sschuberth I have added a comment to the commit message.

Reply:
Commit message typo: s/that it's value field/that its value field/

Reply:
@mnonnenmacher that's what I imagined as well, it requires now 2 ìf`s instead of one. Anyhow...no objections from my side.

Reply:
@fviernau You don't *have* to use the second `if` (i.e. in contrast to `when` they do not have to be exhaustive), you can also do

    if (result is Failure) return emptyList()
    // Do more on success.

Reply:
Hm, if I don't do the second if then I would need to do a `cast` in order to access the properties of `Success` do I? 

If so then then the `!!` required in the previous approach was traded against either a `cast` in the current one. Does this make sense?

Reply:
> If so then then the `!!` required in the previous approach was traded against either a `cast` in the current one.

Correct.

Reply:
Cool so we're talking about the same thing then :-), thanks @sschuberth and @mnonnenmacher for the clarifications!

Reply:
I have used this pattern for early return to avoid the cast in some places which I find nice, because it encapsulates result handling and you usually want to work with the result object only once, afterwards only the value is interesting.

```
val value = when (result) {
  is Success -> result.result
  is Failure -> return whatever
}
```

Reply:
Btw, it's also safer, because in the previous approach with `!!` the value could be set to null on success which would cause runtime errors.

Reply:
> Commit message typo: s/that it's value field/that its value field/

Done.